girder/models/user.py

```diff
@@ -244,26 +249,43 @@ def createUser(self, login, password, firstName, lastName, email,
             'groupInvites': []
         }
 
-        self.setPassword(user, password)
+        self.setPassword(user, password, save=False)
+
+        self.setPublic(user, public, save=False)
+
+        # An id is required in order to grant the user access on itself, but
+        # since the user isn't fully created yet, it should not be saved
+        user['_id'] = ObjectId()
```

Comment:
@zachmullen A bit, but not enough.

My initial concern was actually with ObjectId timestamps being off if the database was running on another machine, but if you go one level deeper into ``save``, the [``_id`` is actually being set](https://github.com/mongodb/mongo-python-driver/blob/2.8/pymongo/collection.py#L385) in the PyMongo client anyway. It's the same in v3.1 (the repo you linked in a really old fork of PyMongo).

The bigger problem, which I initially missed, is indeed that the ``collection.update`` will likely do nothing, since there's no matching document with that ``_id``. Reimplementing the entire ``Model.save`` functionality right here just to explicitly call ``collection.insert`` instead is bad design. I'm also not thrilled with the idea of moving the ``User.setUserAccess`` into a callback, since we don't currently support the case of a user with no access to themselves, so you wouldn't ever want that code to not run.

Any ideas?